### ISSUE TITLE
Broken drawer view sizing when "UINavigationController" with prefers "LargeTitle" is the "DrawerViewController"

### ISSUE THREAD

**Issue by matthewweldon** [52inc/Pulley#214]
Love this pod and the great delegates, just hoping to get some clarity on how this should be handled, not using largeTitles is very incongruous with our app style, and a lot of polished UX apps use them now so I'm hoping it's solvable. 

I've only done a little debugging into this issue so I'm not 100% confident, but I'm reasonably sure the issue is caused by the below scenario. Searched for similar issues but couldn't find something. 

lets say `drawerVC` = my `UINavigationController` that is currently showing a view controller with a long scrollview in its view, and `drawerVC` is inserted as the `drawerContentViewController` in a programmatically instantiated `PulleyViewController`

if: `drawerVC` has `navigationBar.prefersLargeTitles = true` which makes the navigation bar have a larger title and taller initial height

then: when the scroll view is at the top with large title state visible, moving the pulleyDrawer up and down with my finger works fine

**but**: if I scroll the `drawerVC`'s scrollview a little up, making the title in the `navigationBar` transition to the non-large style, and then I use my finger to move the drawer view down and up, the height of the drawer leaves the top safe area when in the `.open` state.

The height the drawer is off by looks to be the same as the difference of the two navigation bar heights

I confirmed that this was the issue by setting `navigationBar.prefersLargeTitles = false` and the drawer heights are perfect again

This happened on both iPhone 7 simulator and iPhone X real device, which I was checked to be sure in case it was safe area handling that was causing the issue.

**Comment by brendan09** [comment]
So, the core issue that you're seeing is that UINavigationController looks for the first scrollview in the hierarchy in order to determine what scrolls / adjusts the large title. This isn't something I can do anything about, unfortunately.

My only recommendation is to use UINavigationController as your primaryContentViewController, rather than putting Pulley inside of the UINavigationController. It may not be ideal, but that may be what you have to do due to UINavigationController's method of selecting the scrollview it wants to interact with.

You can also 'fake' the large title style yourself fairly easily, but that's a bit more work. Sorry I don't have a better answer!

**Comment by matthewweldon** [comment]
Getting back to you a little late, but I have some pics of the problem that provide a little clarity to what I was asking. I suspect you're right about the navigationController, but I wanted to be sure:

Scenario 1, Bottom drawer content is scrolled to the top, large title visible 
Scenario 1, drawer down
<img src="https://user-images.githubusercontent.com/1388193/41791614-42bb5b9a-7624-11e8-9958-045ad54d72db.png" width="100">

Scenario 1, drawer up, safe area interfered with
<img src="https://user-images.githubusercontent.com/1388193/41791629-4d9eba16-7624-11e8-990e-d43c27959031.png" width="100">

Scenario 2, Bottom drawer content is scrolled down a bit, large title no longer visible 
Scenario 2, drawer down
<img src="https://user-images.githubusercontent.com/1388193/41791726-9f8f097a-7624-11e8-977b-49a6917207f7.png" width="100">

Scenario 2, drawer up, safe area is getting interfered with
<img src="https://user-images.githubusercontent.com/1388193/41792119-c16060f2-7625-11e8-9c38-ef33d19a251c.png" width="100">


I think whats happening is the `drawerContentViewController`'s `.opened`height is calculated with the large title view, but when that view shrinks there is some value not being recalculated.

**Comment by brendan09** [comment]
Pulley doesn’t use the contents of the drawer view controller to calculate the height. As a result, that wouldn’t be the case. 

UINavigationController does some weird things with its bounds property / drawing out of bounds in order to accomplish its effects. I don’t know for sure, but I would imagine some internal implementation detail in UINavigationController is causing your issue.

**Comment by matthewweldon** [comment]
that makes sense, with that in mind I may try to handle the bar manually.

**Comment by matthewweldon** [comment]
got back to this a little today, found a work around for now

**Solution:**
setting `collapsedDrawerHeight` and `partialRevealHeight` to constant values in the implementation of `PulleyDrawerViewControllerDelegate` (so ignoring bottom safe area) fixed the visual bugs up top.

**Potential cause:**
from what I can tell, the issue arises when the `setContentOffset` call to the scrollview in the `setDrawerPosition` method uses the lowest available position height (`lowestStop`) of the drawers to calculate scroll offset. 

The calculation of `lowestStop` and the scroll offset made sense, but for some reason the navigation bar height change in the drawer content is able to affect the safe area of the parent container view (the `PulleyViewController.view` in this case)

so the computed var `pulleySafeAreaInsets` in `PulleyViewController.swift` was giving different values depending on the state of the navigation bar, and therefore affected `lowestStop`

doesn't really make sense to me because the `UINavigationController` doesn't contain that view, but I'm chalking that up to uikit magic until i can dig deeper 


**additionally:**
also in my search for answers I have an idea for an optional refactor:

Make a computed var for the `.open` state height that looked like this:
 
```
   private var heightOfOpenDrawer:CGFloat{
        var safeAreaTopInset: CGFloat = 20.0`
        
        if #available(iOS 11.0, *)
        {
            safeAreaTopInset = view.safeAreaInsets.top
        }
        
        return (self.view.bounds.height - topInset - safeAreaTopInset)
        
    }

```

using that in place of either `self.drawerScrollView.bounds.height` or `self.view.bounds.height - topInset - safeAreaTopInset` is a little more semantically accurate: the `.open` height is based on the height of the whole view, not the drawer content, so keeping it referencing this proposed computed var will ignore any unexpected changes in `drawerScrollview` size, if that's even possible.

full disclosure: I don't normally contribute to github so I wasn't sure if it was appropriate to start a PR myself, and I haven't tested my ideas fully, just sharing thoughts. Also I haven't tested for iPad layout.

@brendan09 Thanks for your quick responses, your tip to look into drawing bounds led me to the right solution.

**Comment by brendan09** [comment]
Good call on a computed var for the open state. The inline math ended up getting out of hand with all of the iterations on computing height.

I merged that in and it's in version 2.4.3 (now on CocoaPods). Thanks!
